### PR TITLE
Update Syn

### PR DESCRIPTION
**Pull request by nrc**
Also fixes a compile issue with the `nightly` feature and updates procinfo to use the same fork as TiKV.

PTAL @breeswish  @Hoverbear

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -35,7 +35,7 @@ spin = "0.5"
 reqwest = { version = "0.9.5", optional = true }
 
 [target.'cfg(target_os = "linux")'.dependencies]
-procinfo = { version = "0.4", optional = true }
+procinfo = { git = "https://github.com/tikv/procinfo-rs", optional = true }
```
Comment (breeswish) [inline Cargo.toml]: maybe not appropriate to use pingcap's own procinfo?

File: static-metric/Cargo.toml [hunk 2]
```diff
@@ -17,9 +17,9 @@ include = [
 proc-macro = true
 
 [dependencies]
-syn = { version = "0.13", features = ["full", "extra-traits"] }
-proc-macro2 = "0.3"
-quote = "0.5"
+syn = { version = "0.15", features = ["full", "extra-traits"] }
```
Comment (breeswish) [inline static-metric/Cargo.toml]: syn has version 1.0.3: https://docs.rs/syn/1.0.3/syn/

File: static-metric/Cargo.toml [hunk 3]
```diff
@@ -17,9 +17,9 @@ include = [
...
+proc-macro2 = "0.4"
+quote = "0.6"
```
Comment (breeswish) [inline static-metric/Cargo.toml]: quote has version 1.0.2: https://docs.rs/quote/1.0.2/quote/

File: static-metric/Cargo.toml [hunk 4]
```diff
@@ -17,9 +17,9 @@ include = [
 proc-macro = true
 
 [dependencies]
-syn = { version = "0.13", features = ["full", "extra-traits"] }
-proc-macro2 = "0.3"
-quote = "0.5"
+syn = { version = "0.15", features = ["full", "extra-traits"] }
+proc-macro2 = "0.4"
```
Comment (breeswish) [inline static-metric/Cargo.toml]: proc-macro2 has version 1.0.1: https://docs.rs/proc-macro2/1.0.1/proc_macro2/

File: Cargo.toml (hunk 1 above)
Comment (nrc) [reply Cargo.toml]: Why not? We use it in TiKV

Comment (lucab) [reply Cargo.toml]: But `tikv/procinfo-rs` is not published as a crate, right? Which would make this crate also not publishable anymore.
I do personally have projects (already on crates.io) which need to consume this from the registry.

Have you maybe considered switching to https://docs.rs/procfs/0.5.3/procfs/ or checking with @danburkert if they are open to co-maintainers? (does PingCAP want to maintain that crate?)

File: static-metric/Cargo.toml (hunk 2 above)
Comment (kennytm) [reply static-metric/Cargo.toml]: IMO we should update TiKV's `syn` to 1.0 too.

File: Cargo.toml (hunk 1 above)
Comment (nrc) [reply Cargo.toml]: I can change this back if we think it is really important, the key thing is that we move off 0.3 which the version of prometheus in TiKV is currently using

File: static-metric/Cargo.toml (hunk 2 above)
Comment (nrc) [reply static-metric/Cargo.toml]: The issue is not what version TiKV is using, but what version it's deps are using (Serde, Prost, slog, structopt, and more). We *should* send PRs to update them all, but I don't want to take that on right now when there is low-hanging fruit in just updating Prometheus to use a more recent version. Let's not let perfect be the enemy of good here.

Comment (nrc) [reply static-metric/Cargo.toml]: ping @breeswish

Comment (breeswish) [reply static-metric/Cargo.toml]: Ok

Comment (breeswish) [reply static-metric/Cargo.toml]: But what's the purpose of updating these versions?

**Comment by nrc** [comment]
ping @Hoverbear @breeswish changed procinfo to version 4 so we can still publish the crate. PTAL.

File: static-metric/src/parser.rs [hunk 5]
```diff
@@ -14,28 +14,23 @@
...
-impl Synom for LabelEnum {
-    fn parse(tokens: Cursor) -> PResult<Self> {
-        if let Some((term, rest)) = tokens.term() {
-            if term.as_str() == "label_enum" {
-                return Ok((LabelEnum { span: term.span() }, rest));
-            }
+impl Parse for LabelEnum {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let ident: Ident = input.parse()?;
+        if &ident.to_string() == "label_enum" {
```
Comment (Hoverbear) [inline static-metric/src/parser.rs]: Is the `.to_string()` here an unnecessary copy?

Comment (nrc) [reply static-metric/src/parser.rs]: It is a copy, I'm not sure if it is unnecessary - I don't think there is a better way of getting a string repr of an `Ident`, and turning a string into an ident also requires a copy. It also doesn't matter - the ident name will always be short and this code only executes at compile time and is not hot.

**Comment by breeswish** [comment]
LGTM. I prefer to add testing script of the static-metric to the travis as well (e.g. building the library as well as its examples) to make sure that everything is ok.

File: static-metric/src/parser.rs (hunk 5 above)
Comment (kennytm) [reply static-metric/src/parser.rs]: @nrc `syn 0.15.44` has [this impl](https://docs.rs/syn/0.15.44/syn/struct.Ident.html#impl-PartialEq%3CT%3E):

```rust
impl<T> PartialEq<T> for Ident where
    T: AsRef<str> + ?Sized, 
```

so `ident == "label_enum"` (add or minus a few `&` and `*`) should just work.

**Comment by nrc** [comment]
ping @Hoverbear @breeswish PTAL

**Comment by breeswish** [comment]
How about the https://github.com/pingcap/rust-prometheus/pull/268#issuecomment-528170310 ?

**Comment by nrc** [comment]
> How about the #268 (comment) ?

@breeswish  Its is addressed in https://github.com/pingcap/rust-prometheus/pull/268/files#diff-354f30a63fb0907d4ad57269548329e3, I think

**Comment by nrc** [comment]
@kennytm PTAL (I've addressed @Hoverbear's and your comments)

**Pull request merged**